pn.1
.TH PN 1
.SH NAME

pn - a command-line tool for phone number manipulation

.SH SYNOPSIS

.B pn
.I COMMAND
[\fIOPTIONS\fR]
.I INPUT

.SH COMMANDS

.TP
.B valid
Tests whether the phone number given as \fIINPUT\fR is a valid number or short
number. Returns \fB0\fR (SUCCESS) if the number is valid, \fB1\fR (ERROR)
otherwise. Valid options: \fB-c\fR, \fB-v\fR (verbose, default off).
.TP
.B info
Displays informations about the phone number given as \fIINPUT\fR. Valid
option: \fB-c\fR.
.TP
.B format
Formats a telephone number given as \fIINPUT\fR to the specified standard
(E164 by default), result on stdout. Valid options: \fB-c\fR, \fB-f\fR.
.TP
.B find
Finds phone number in some text given as \fIINPUT\fR, result on stdout (one
line per number found).  Valid options: \fB-c\fR, \fB-f\fR, \fB-l\fR.
.TP
.B dialout
Formats a phone number given as \fIINPUT\fR in \fBe164\fR format for
out-of-country dialing purposes. Valid options: \fB-c\fR, \fB-l\fR.

.SH OPTIONS

.TP
\fB-c\fR \fICOUNTRY\-CODE\fR
If set, the phone number(s) given in \fIINPUT\fR will be considered as being
dialed from the country specified by the 2 uppercase letters
\fICOUNTRY\-CODE\fR (e.g.  "FR", "US", ...). If this option is not set,
\fBpn\fR will look for numbers in the international format.
.TP
\fB-f\fR \fIFORMAT\fR
Sets the phone number output format. \fIFORMAT\fR must be one of \fBe164\fR,
\fBint\fR (international), \fBnat\fR (national), or \fBteluri\fR (RFC3966 tel
URI). The default value is \fBe164\fR.
.TP
\fB-l\fR \fILENIENCY\fR
Sets the phone number search leniency in \fIINPUT\fR. \fILENIENCY\fR must be
one of:
.RS
.TP
.B possible
Phone numbers accepted are possible, but not necessarily valid.
.TP
.B valid
(default value) Phone numbers accepted are possible and valid. Numbers written
in national format must have their national-prefix present if it is usually
written for a number of this type.
.TP
.B strict
Phone numbers accepted are valid and are grouped in a possible way for this
locale. For example, a US number written as "65 02 53 00 00" and "650253 0000"
are not accepted at this leniency level, whereas "650 253 0000", "650 2530000"
or "6502530000" are. Numbers with more than one '/' symbol in the national
significant number are also dropped at this level.
.TP
.B exact
Phone numbers accepted are valid and are grouped in the same way that we would
have formatted it, or as a single block. For example, a US number written as
"650 2530000" is not accepted at this leniency level, whereas "650 253 0000" or
"6502530000" are. Numbers with more than one '/' symbol are also dropped at
this level.

(see https://javadoc.io/doc/com.googlecode.libphonenumber/libphonenumber/8.4.1
for more details)
.RE

.SH RETURN STATUS
\fBpn\fR returns 0 on successful operations, 1 if an error occurred (option
error, parsing error).

.SH AUTHOR
Camille Oudot <camille.oudot@orange.com>


src/pn.cc
/*
 * pn - a command-line tool for phone number manipulation
 *
 * Copyright 2017 Orange
 * <camille.oudot@orange.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <iostream>
#include <getopt.h>
#include <string.h>
#include <limits>
#include <phonenumbers/phonenumbermatcher.h>
#include <phonenumbers/phonenumberutil.h>
#include <phonenumbers/phonenumbermatch.h>
#include <phonenumbers/shortnumberinfo.h>
#include <phonenumbers/phonemetadata.pb.h>
#include <phonenumbers/geocoding/phonenumber_offline_geocoder.h>

using namespace std;
using namespace i18n::phonenumbers;

string country_code = "ZZ";
enum PhoneNumberUtil::PhoneNumberFormat number_format = PhoneNumberUtil::E164;
enum PhoneNumberMatcher::Leniency matcher_leniency = PhoneNumberMatcher::VALID;


bool parse_format(char *s, enum PhoneNumberUtil::PhoneNumberFormat *f)
{
	if (strcmp(s, "e164") == 0) {
		*f = PhoneNumberUtil::E164;
	} else if (strcmp(s, "int") == 0) {
		*f = PhoneNumberUtil::INTERNATIONAL;
	} else if (strcmp(s, "nat") == 0) {
		*f = PhoneNumberUtil::NATIONAL;
	} else if (strcmp(s, "teluri") == 0) {
		*f = PhoneNumberUtil::RFC3966;
	} else {
		return false;
	}
	return true;
}

bool parse_leniency(char *s, enum PhoneNumberMatcher::Leniency *l)
{
	if (strcmp(s, "possible") == 0) {
		*l = PhoneNumberMatcher::POSSIBLE;
	} else if (strcmp(s, "valid") == 0) {
		*l = PhoneNumberMatcher::VALID;
	} else if (strcmp(s, "strict") == 0) {
		*l = PhoneNumberMatcher::STRICT_GROUPING;
	} else if (strcmp(s, "exact") == 0) {
		*l = PhoneNumberMatcher::EXACT_GROUPING;
	} else {
		return false;
	}
	return true;
}

int format(int argc, char *argv[])
{
	int c;
	while ((c = getopt(argc, argv, "c:f:")) != -1) {
		switch((char)c) {
		case 'c':
			country_code = optarg;
			break;
		case 'f':
			if (!parse_format(optarg, &number_format)) {
				cerr << "invalid number format: " << optarg << endl;
				return 1;
			}
			break;
		default:
			return 1;
			break;
		}
	}

	if (optind == argc) {
		cerr << "missing the phone number as argument" << endl;
		return 1;
	}

	PhoneNumberUtil *pnu = PhoneNumberUtil::GetInstance();
	PhoneNumber pn;
	if (pnu->Parse(argv[optind], country_code, &pn) != PhoneNumberUtil::NO_PARSING_ERROR) {
		cerr << "could not parse number: " << argv[optind] << endl;
		return 1;
	}

	string s;
	pnu->Format(pn, number_format, &s);
	std::cout << s << std::endl;

	return 0;
}

int find(int argc, char *argv[])
{
	int c;
	while ((c = getopt(argc, argv, "c:f:l:")) != -1) {
		switch((char)c) {
		case 'c':
			country_code = optarg;
			break;
		case 'f':
			if (!parse_format(optarg, &number_format)) {
				cerr << "invalid number format: " << optarg << endl;
				return 1;
			}
			break;
		case 'l':
			if (!parse_leniency(optarg, &matcher_leniency)) {
				cerr << "invalid leninency: " << optarg << endl;
				return 1;
			}
			break;
		default:
			return 1;
			break;
		}
	}

	if (optind == argc) {
		cerr << "missing the text to parse as argument" << endl;
		return 1;
	}

	PhoneNumberUtil *pnu = PhoneNumberUtil::GetInstance();
	PhoneNumberMatcher pnm(*pnu, argv[optind], country_code, matcher_leniency, numeric_limits<int>::max());
	PhoneNumberMatch match;
	string s;

	while (pnm.HasNext())
	{
		pnm.Next(&match);
		pnu->Format(match.number(), number_format, &s);
		std::cout << s << std::endl;

	}
	return 0;
}

const char *num_type_descr(PhoneNumberUtil::PhoneNumberType pnt)
{
	switch (pnt) {
	case PhoneNumberUtil::FIXED_LINE:
		return "fixed line";
		break;
	case PhoneNumberUtil::MOBILE:
		return "mobile";
		break;
	case PhoneNumberUtil::FIXED_LINE_OR_MOBILE:
		return "fixed line or mobile";
		break;
	case PhoneNumberUtil::TOLL_FREE:
		return "toll free";
		break;
	case PhoneNumberUtil::PREMIUM_RATE:
		return "premium rate";
		break;
	case PhoneNumberUtil::SHARED_COST:
		return "shared cost";
		break;
	case PhoneNumberUtil::VOIP:
		return "voip";
		break;
	case PhoneNumberUtil::PERSONAL_NUMBER:
		return "personal number";
		break;
	case PhoneNumberUtil::PAGER:
		return "pager";
		break;
	case PhoneNumberUtil::UAN:
		return "uan";
		break;
	case PhoneNumberUtil::VOICEMAIL:
		return "voicemail";
		break;
	case PhoneNumberUtil::UNKNOWN:
	default:
		return "unknown";
		break;
	}
}

int info(int argc, char *argv[])
{
	int c;
	while ((c = getopt(argc, argv, "c:")) != -1) {
		switch((char)c) {
		case 'c':
			country_code = optarg;
			break;
		default:
			return 1;
			break;
		}
	}

	if (optind == argc) {
		cerr << "missing the phone number as argument" << endl;
		return 1;
	}

	PhoneNumberUtil *pnu = PhoneNumberUtil::GetInstance();
	PhoneNumber pn;
	const string raw_num = argv[optind];

	if (pnu->Parse(raw_num, country_code, &pn) != PhoneNumberUtil::NO_PARSING_ERROR) {
		cerr << "could not parse number: " << argv[optind] << endl;
		return 1;
	}

	const ShortNumberInfo sni;
	const PhoneNumberOfflineGeocoder pnog;
	const Locale loc("C");
	string s;

	pnu->GetRegionCodeForCountryCode(pn.country_code(), &s);
	cout << "country code: " << pn.country_code() << " (" << s << ")" << endl;
	cout << "number type: " << num_type_descr(pnu->GetNumberType(pn)) << endl;
	cout << "location: " << pnog.GetDescriptionForNumber(pn, loc) << endl;
	cout << "possible short number: " << (sni.IsPossibleShortNumber(pn) ? "true" : "false") << endl;
	cout << "valid short number: " << (sni.IsValidShortNumber(pn) ? "true" : "false") << endl;
	cout << "emergency number: "<< (sni.IsEmergencyNumber(raw_num, country_code) ? "true" : "false") << endl;

	return 0;
}

int valid(int argc, char *argv[])
{
	int c, ret;
	bool verbose = false;
	string msg;
	while ((c = getopt(argc, argv, "c:v")) != -1) {
		switch((char)c) {
		case 'c':
			country_code = optarg;
			break;
		case 'v':
			verbose = true;
			break;
		default:
			return 1;
			break;
		}
	}

	if (optind == argc) {
		cerr << "missing the phone number as argument" << endl;
		return 1;
	}

	PhoneNumberUtil *pnu = PhoneNumberUtil::GetInstance();
	PhoneNumber pn;
	const string raw_num = argv[optind];

	if (pnu->Parse(raw_num, country_code, &pn) != PhoneNumberUtil::NO_PARSING_ERROR) {
		switch (pnu->Parse(raw_num, country_code, &pn)) {
		case PhoneNumberUtil::INVALID_COUNTRY_CODE_ERROR:
			msg = "invalid country code";
			break;
		case PhoneNumberUtil::NOT_A_NUMBER:
			msg = "not a number";
			break;
		case PhoneNumberUtil::TOO_SHORT_AFTER_IDD:
			msg = "too short after international dialing prefix";
			break;
		case PhoneNumberUtil::TOO_SHORT_NSN:
			msg = "too short";
			break;
		case PhoneNumberUtil::TOO_LONG_NSN:
			msg = "too long";
			break;
		default:
			msg = "unknown error";
			break;
		}
		if (verbose) {
			cout << msg << endl;
		}
		return 1;
	}

	const ShortNumberInfo sni;

	if (!pnu->IsValidNumber(pn)) {
		if (sni.IsValidShortNumber(pn)) {
			if (verbose) {
				cout << "valid short number" << endl;
			}
			return 0;
		}
		if (verbose) {
			cout << "invalid number" << endl;
		}
		return 1;
	}

	if (verbose) {
		cout << "valid number" << endl;
	}
	return 0;
}

int dialout(int argc, char *argv[])
{
	int c, ret;
	bool verbose = false;
	string msg;
	while ((c = getopt(argc, argv, "c:v")) != -1) {
		switch((char)c) {
		case 'c':
			country_code = optarg;
			break;
		case 'v':
			verbose = true;
			break;
		default:
			return 1;
			break;
		}
	}

	if (optind == argc) {
		cerr << "missing the phone number as argument" << endl;
		return 1;
	}

	PhoneNumberUtil *pnu = PhoneNumberUtil::GetInstance();
	PhoneNumber pn;
	const string raw_num = argv[optind];

	if (pnu->Parse(raw_num, "ZZ", &pn) != PhoneNumberUtil::NO_PARSING_ERROR) {
		cerr << "invalid input number " << raw_num << " (must a valid number in e164 format, e.g. +124567890)" << endl;
		return 1;
	}

	string result;

	pnu->FormatOutOfCountryCallingNumber(pn, country_code, &result);

	if (result == "") {
		cerr << "couldn't determine the dial out sequence for " << raw_num << endl;
	}

	cout << result << endl;
	return 0;
}

int main(int argc, char *argv[])
{
	if (argc < 3) {
		// TODO usage
		return 1;
	}

	if (strcmp(argv[1], "format") == 0) {
		return format(argc - 1, &argv[1]);
	} else if (strcmp(argv[1], "find") == 0) {
		return find(argc - 1, &argv[1]);
	} else if (strcmp(argv[1], "info") == 0) {
		return info(argc - 1, &argv[1]);
	} else if (strcmp(argv[1], "valid") == 0) {
		return valid(argc - 1, &argv[1]);
	} else if (strcmp(argv[1], "dialout") == 0) {
		return dialout(argc - 1, &argv[1]);
	} else if (strcmp(argv[1], "-h") == 0) {
		//usage();
	} else {
		cerr << "invalid command: " << argv[1] << endl;
		return 1;
	}
}
